### ISSUE TITLE
CSP can be circumvented in HTML5 using obsolete <font>

### ISSUE THREAD

**Issue by mattok** [MoonchildProductions/Pale-Moon#779]
This is [fundamentally a problem with HTML itself, which has already been reported to w3c](https://github.com/w3c/html/issues/735). However, since this affects Pale Moon too, and since it will probably take some time before the issue gets addressed in HTML by W3C, I thought I'd report it here too. The basic behaviour I'd expect is that if CSP has been used to block inline styles, then `<font>` elements should be ignored (unless accompanied by a valid CSP nonce, perhaps).

Aside: Within HTML5 documents, I wonder whether there are other obsoleted elements that require attention too from the point of view of CSP too...

**Comment by wolfbeast** [comment]
I think if web designers are going to use CSP to address XSS concerns, that they should most definitely not still be using obsolete tags in their web pages and mixing HTML specs/versions.

I'm also not sure what you expect to be done on the browser side here -- in-line replacing of HTML tags is not a good idea and can cause all sorts of unexpected results. I also don't think this responsibility should be off-loaded onto the web clients.

**Comment by mattok** [comment]
> I think if web designers are going to use CSP to address XSS concerns, that they should most definitely not still be using obsolete tags in their web pages and mixing HTML specs/versions.

This doesn't make sense. CSP allows you to control what third parties can do, and, for example, protect against potentially harmful consequences should a third party manage to inject some markup into your document (through some otherwise undiscovered SQL injection vulnerability or whatever). Unfortunately, HTML5 has no 'strict' mode as far as I am aware, and as such, permits rendering of obsolete elements. So, even if a web designer writes perfect valid HTML5 code, without  obsolete elements, a malicious actor can still make use of obsolete (and potentially insecure) HTML elements... The only protection I know against this is CSP ... but if not even CSP works, we're in a bad place. I think this is primarily a problem with HTML, hence why I reported it to the W3C first. However, if there is something that can be done in Pale Moon now, to make it more secure than HTML itself requires, that would be great.

> I'm also not sure what you expect to be done on the browser side here -- in-line replacing of HTML tags is not a good idea and can cause all sorts of unexpected results. I also don't think this responsibility should be off-loaded onto the web clients.

CSP **is** the responsibility of the web-cllents, but CSP is probably ambiguous about how to interpret `font` tags.

I can tell you what effect I would expect a CSP rule that blocks inline styles to have on a document with `<font>` elements in it: The `font` elements will effectively be ignored, and the document will render as if they weren't present. However, I can't/wont tell you _how_ Pale Moon should achieve that. 

I don't know how Pale Moon or other browsers actually implement CSP. It presume that HTML is not simply edited to sanitise it according to CSP rules before processing, because that would not protect against DOM manipulation through javascript, etc. So, I'm guessing that CSP rules take effect either at the DOM level - preventing certain elements from being inserted, or at a later stage, presenting elements that have been inserted but which don't meet certain rules from having any effect.

If anyone has got the time/inclination to look at this (perhaps not just for the specific case of `font` elements, but also for other potential obsoleted script and style manipulating/accepting elements or attributes), I for one would be very grateful.

**Comment by wolfbeast** [comment]
CSP is there so third parties cannot inject code in authored documents in the first place...

It's not up to the client to enforce what kind of (otherwise valid) tags are allowed or not in a document.

It's possible I misunderstood what the issue is here. Can you give an example where a **third party** can use the `<font>` tag's in-line styles in a malicious way?

**Comment by mattok** [comment]
> CSP is there so third parties cannot inject code in authored documents in the first place...

That's not how I see it. It is also to _mitigate the risk of content injection_ should it occur. Preventing third parties from injecting code in the first place is one way to achieve that. [From the standard](https://www.w3.org/TR/CSP3/#intro):

> This document defines Content Security Policy (CSP), a tool which developers can use to lock down their applications in various ways, mitigating the risk of content injection vulnerabilities such as cross-site scripting, and reducing the privilege with which their applications execute.

> CSP is not intended as a first line of defense against content injection vulnerabilities. Instead, CSP is best used as defense-in-depth. It reduces the harm that a malicious injection can cause, but it is not a replacement for careful input validation and output encoding.

So, it's not just about blocking the loading of resources from other domains, for example. It's also supposed to offer protection against, for example, specially crafted HTTP requests (GET/POST/whatever) resulting in the application itself displaying unintended content.

> It's not up to the client to enforce what kind of (otherwise valid) tags are allowed or not in a document.

This isn't about tags, as I see it. It's about behaviour. If the CSP rule says, "nothing in the HTML content should be interpreted as a request to change the style of the document" or "nothing in the HTML content should be interpreted as a script to be run", then any such markup, no matter what it is or how it got there, should have zero effect.

> Can you give an example where a third party can use the `<font>` tag's in-line styles in a malicious way?

I don't think this is a completely valid point. This is supposed to be **defense-in-depth**. So, I shouldn't need to say what font- or browser- related vulnerability someone may targetting _now_. The idea is that the possibility of someone using it maliciously at any time is eliminated.

However, it is worth considering how it could possibly be used maliciously now. The `font` element can be officially used to set the font colour, weight or size, but in the past various browsers, including Firefox, allowed non-standard font attributes to be set too, such as font weight and point size. I don't specifically know how it could be used maliciously, but here are the kinds of things I think someone might try to achieve if they could insert it:
- Use the styling to hide some content they had injected.
- Attempt to target some older, non-official behaviour, previously supported and still available in the browser. (I don't know if Pale Moon has any such thing.)
- Attempt to load a font-face known to be vulnerable (possibly, because the actor know that it was already available on the system, or because they had managed to upload the font via some other means, etc).

Taking the last point, a quick search for, for example, "Firefox font-face vulnerability" shows that such vulnerabilities have existed in an exploitable way in the past. So, although these are hypothetical ways the `font` could be used maliciously, they are not completely without basis.

One of the things that draws me to Pale Moon is the defense-in-depth security enhancements. If it was decided to implement this at some point in the future too, then I personally would be very happy about it.

**Comment by wolfbeast** [comment]
> CSP is not intended as a first line of defense against content injection vulnerabilities. Instead, CSP is best used as defense-in-depth. It reduces the harm that a malicious injection can cause, but it is not a replacement for careful input validation and output encoding.

I think you miss the point here. The point is that CSP is a catch-all in case someone isn't using validation of passed variables. That is your defense-in-depth, because server-side code like PHP may have potential XSS vulnerabilities. In-line restrictions are mostly for reflected XSS -- given that, I don't see how a `<font>` tag would ever pose a threat (you can only refer to an already-defined font, not set one).

I'll keep it in mind, though. As for an actual implementation to go along with this, I'm not sure. I gather simply dropping the tag params on the floor would be enough. `<font>some text</font>` after all doesn't do anything.

**Comment by mattok** [comment]
I'm not sure I agree with inline restrictions being for reflected XSS only (or mostly), but perhaps I missed something.

Please note that while the deprecated `<font>` element is the first one I discovered a potential issue with, and therefore reported, there could be other more serious issues.

I've just been taking a look through other deprecated/obsoleted elements and attributes that could pose a security risk.

There are a couple ( e.g. `<center>`, `<strike>`), that are similar to `<font>` in that they have been obsoleted and they affect the style only. However,  they don't load resources and can't be used to hide content like `<font>` potentially could. The chances of someone exploiting a vulnerability through them are slim to say the least. Also, from your description, it appears that applying CSP to them would be more difficult than for `<font>`, because you'd have to remove the elements themselves, or replace them by something else, like `<span>`. So, not worth bothering about those.

I did notice that Pale Moon still renders the obsoleted `<marquee>` element though, and furthermore, that it supports three event handlers, `onbounce`, `onstart` and `onfinish`. I did some testing, and the good news is that Pale Moon already applies CSP rules to these event handlers. If a `default-src: 'self'` is set, then they are blocked. The odd thing is that I seemed to have to set `'unsafe-eval'` to unblock them, rather than `'unsafe-inline'` as I would have expected. To reproduce this, do the following in PaleMoon 27.0.3:

1. Save the following to index.php
```php
<?php header("Content-Security-Policy: default-src 'self'; script-src 'self' 'unsafe-inline'"); ?><!doctype html>
<html>
<head><title>marquee test</title></head>
<body><marquee onbounce="console.log('bounced');" onfinish="console.log('finished');" onstart="console.log('started');" direction="down" width="250" height="200" behavior="alternate" style="border:solid">
  <marquee behavior="alternate">
    This text will bounce
  </marquee>
</marquee>
</body>
</html>
```
2. From the same directory as index.php, run `php -S localhost:8080 index.php`.

3. Access [](http://localhost:8080/index.php) in your browser, and open the developer tools to check what's happening in the console. Note that the events aren't firing.

4. Modify index.php to add `'unsafe-eval'` after`'unsafe-inline'` in the CSP header.

5. Refresh the index.php page in the browser. Note that the events start to fire, producing messages in the console log.

This is probably a bug in the CSP implementation that you may wish to take a look at.

**Comment by wolfbeast** [comment]
I think unsafe-eval is correct there. They are not in-line scripts, but rather eval()s of the handler code.

**Comment by wolfbeast** [comment]
Also, as far as "marquee" and other obsolete elements are concerned: unless specified with a doctype, the rendering mode for HTML is an amalgamation to render "as much as possible". This is the case for all browsers; to have maximum web compatibility. Marquee is also still rendered by Firefox, for example, and that will likely not go away any time soon -- in fact, the most recent update of Pale Moon has a security update for Marquee tags.

I'm all for dropping marquee, actually ;) I hate the damn thing and it's got a whole bunch of special snowflake treatments in the Mozilla code.

**Comment by wolfbeast** [comment]
Dveditz from MozSec echoed my lack of concern for abuse here:

> I'm not too worried about this because it doesn't allow for XSS or bypassing content-loading restrictions which are the main thrust of CSP. It's a tiny part of what <style> can do, but other than making some text disappear or fill up the page it can't move content around or other transformations that style can do.

So, the advise remains to let site authors know about it in case they need to protect themselves on specific pages. A better forum for this would be an issue on the CSP spec in github: https://github.com/w3c/webappsec-csp/issues

**Comment by mattok** [comment]
> I think unsafe-eval is correct there. They are not in-line scripts, but rather eval()s of the handler code.

[According to Mozilla on CSP](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/script-src),

> **'unsafe-inline'**
> Allows the use of inline resources, such as inline <script> elements, javascript: URLs, inline event handlers, and inline <style> elements. You must include the single quotes.
> **'unsafe-eval'**
> Allows the use of eval() and similar methods for creating code from strings. You must include the single quotes.

I interpret `onbounce`, `onfinish` and `onstart` as inline event handlers, which I would expect to be covered by `unsafe-inline`. There are no explicit `eval`s in the example code I provided. I think you're saying that behind the scenes within Pale Moon, the javascript within the event handler is implicitly interpreted as a string and `eval`'ed, and so CSP blocks it unless you do `unsafe-eval` too. Is that right?

The issue I have with that, is, given some legacy code that needs managing and which contains javascript within inline event handlers (I've got some like that), it's not possible to use CSP to say "allow the legacy event handlers, but don't allow `eval` to be run anywhere - including explicitly within the event handlers". You've either got to allow unsafe inline event handlers and allow unsafe eval to be used, or not permit inline event handlers at all. So, it feels to me that this interpretation of CSP is less secure for legacy code, simply because it's less flexible in terms of what you can and cannot allow. Your interpretation may well be correct, but it wasn't what I was expecting given the documentation I've read.

**Comment by wolfbeast** [comment]
You're not correct for the unsafe-inline. We're not talking about event handlers here, we're talking about a (small) subset of style rules that can be applied in the legacy way that has neither anything to do with on* handlers (which are JS) nor with in-line `<style>` elements.  So it falls between the cracks either way.

The problem seems to be you want CSP to do everything. That's not what it does. CSP is meant to be a supportive measure (as you stated yourself, it's a DiD measure) that can not ever replace legacy, bad, or outdated code in the pages themselves.

> You've either got to allow unsafe inline event handlers and allow unsafe eval to be used, or not permit inline event handlers at all. 

Patches are welcome.

**Comment by mattok** [comment]
I think that we're talking at cross purposes a bit here.

Some obsoleted elements (`<font>`, `<center>`, etc.) allow the user to affect styling in a manner that is not allowed in current HTML. CSP doesn't currently apply to the obsoleted elements in the same way that it already applies to other elements. I've argued that it should and as you know, I've submitted an HTML issue to that effect. You've argued that it's a non-issue, because you don't see the setting of styles in this manner as providing an avenue for abuse. I'm not sure, but if that's the case then I'm happy with the status quo. I've never said that I want CSP to do everything.

Some obsoleted elements (`<marquee>`) also have attributes that are inline event handlers accepting javascript (`onbounce`, `onstart` and `onfinish` for the `<marquee>` element). The default CSP rules block inline event handlers and javascript `eval`. The standard says that `'unsafe-line'` can be used to allow the inline event handlers, and that `'unsafe-eval'` can be used to allow javascript `eval()`. However, Pale Moon requires both `'unsafe-inline'` and `'unsafe-eval'` to be set to make the event hander operational - even if the event handler javascript itself doesn't contain an `eval()`. I think this is a bug in Pale Moon's CSP implementation. I only report this here, because I encountered the issue while testing out obsoleted elements as part of the original `<font>` issue.

When I said

> You've either got to allow unsafe inline event handlers and allow unsafe eval to be used, or not permit inline event handlers at all.

I meant that, with Pale Moon as it stands now, someone using CSP must allow both unsafe inline event handlers and allow unsafe eval to be used if they want inline event handlers to operate. I wasn't suggesting that some specific change had to be made in the code. Anyhow, before anyone starts to make a patch (and sorry, but I'm probably not the right person), some agreement on whether there is a bug in the CSP implementation or not needs to be arrived at.
